File: App/Services/Notebook/NotebookRepository.swift
import Foundation
import Supabase

@MainActor final class NotebookRepository {
    private let sdk: SupabaseClient
    init(sdk: SupabaseClient) { self.sdk = sdk }
    private struct ReadPayload: Encodable {
        var note: UUID? = nil
        let after: UUID?
        enum CodingKeys: String, CodingKey { case p_note, p_after }
        func encode(to encoder: Encoder) throws {
            var c = encoder.container(keyedBy: CodingKeys.self)
            if let note { try c.encode(note, forKey: .p_note) } else { try c.encodeNil(forKey: .p_note) }
            if let after { try c.encode(after, forKey: .p_after) } else { try c.encodeNil(forKey: .p_after) }
        }
    }
    lazy var reader = NotebookReader(current: { [weak self] in self?.identity() }, read: { [weak self] after in
        guard let self else { throw NotebookFailure.unavailable }
        return try await self.sdk.rpc("isg_notebook_read_v1", params: ReadPayload(after: after)).execute().data
    })
    func snapshot(_ identity: NotebookIdentity) throws -> NotebookReader.Snapshot {
        try reader.snapshot(identity, drafts: queue.pending(identity))
    }
    func organization(_ note: UUID, identity expected: NotebookIdentity) async throws -> NotebookOrganization {
        guard identity() == expected else { throw NotebookFailure.identityChanged }
        struct Params: Encodable { let p_note: UUID }
        let data = try await sdk.rpc("isg_notebook_organization_v1", params: Params(p_note: note)).execute().data
        try Task.checkCancellation()
        guard identity() == expected, data.count <= 2_000_000 else { throw NotebookFailure.identityChanged }
        let value = try JSONDecoder().decode(NotebookOrganization.self, from: data); try value.validate(note); return value
    }
    private struct ReminderReadPayload: Encodable {
        let p_after: UUID?
        func encode(to encoder: Encoder) throws {
            var values = encoder.container(keyedBy: CodingKeys.self)
            if let p_after { try values.encode(p_after, forKey: .p_after) }
            else { try values.encodeNil(forKey: .p_after) }
        }
        enum CodingKeys: String, CodingKey { case p_after }
    }
    func reminders(_ expected: NotebookIdentity) async throws -> [NotebookReminder] {
        guard identity() == expected else { throw NotebookFailure.identityChanged }
        var result: [NotebookReminder] = []
        var after: UUID?
        for _ in 0..<50 {
            let data = try await sdk.rpc("isg_notebook_reminders_v1", params: ReminderReadPayload(p_after: after)).execute().data
            try Task.checkCancellation()
            guard identity() == expected, data.count <= 2_000_000 else { throw NotebookFailure.identityChanged }
            let page = try JSONDecoder().decode(NotebookReminderPage.self, from: data)
            try page.validate(after: after)
            result.append(contentsOf: page.reminders)
            guard result.count <= 1_000 else { throw NotebookFailure.full }
            guard page.has_more else { return result }
            after = page.next_after
        }
        throw NotebookFailure.full
    }
    @discardableResult
    func createReminder(title: String, recurrence: NotebookReminderRecurrence, dueAt: Date,
                        note: UUID? = nil, identity expected: NotebookIdentity) async throws -> UUID {
        guard identity() == expected, dueAt > Date(),
              !title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              title.unicodeScalars.count <= 200 else { throw NotebookFailure.invalid }
        let zone = TimeZone.current
        let day = DateFormatter(); day.calendar = Calendar(identifier: .gregorian); day.locale = Locale(identifier: "en_US_POSIX")
        day.timeZone = zone; day.dateFormat = "yyyy-MM-dd"
        let time = DateFormatter(); time.calendar = Calendar(identifier: .gregorian); time.locale = Locale(identifier: "en_US_POSIX")
        time.timeZone = zone; time.dateFormat = "HH:mm:ss"
        let payload = NotebookReminderMutationPayload(mutation: UUID(), action: "create", reminder: nil,
            note: note, occurrence: nil, expected: 0, title: title, recurrence: recurrence,
            localTime: time.string(from: dueAt), startsOn: day.string(from: dueAt), timezone: zone.identifier,
            snoozedUntil: nil, installation: NotificationService.shared.serverPushInstallationID)
        let receipt = try await reminderMutation(payload, identity: expected)
        guard receipt.state == "active", receipt.delivery_strategy == "server_push",
              let reminder = receipt.reminder_id else { throw NotebookFailure.invalid }
        return reminder
    }
    func settleReminder(_ action: String, reminder: NotebookReminder,
                        occurrence: NotebookReminderOccurrence? = nil, snoozedUntil: Date? = nil,
                        identity expected: NotebookIdentity) async throws {
        guard ["complete", "snooze", "cancel"].contains(action), reminder.state == "active",
              (action == "cancel") == (occurrence == nil),
              (action == "snooze") == (snoozedUntil != nil) else { throw NotebookFailure.invalid }
        let payload = NotebookReminderMutationPayload(mutation: UUID(), action: action, reminder: reminder.reminder_id,
            note: nil, occurrence: occurrence?.occurrence_id, expected: reminder.series_version, title: nil,
            recurrence: nil, localTime: nil, startsOn: nil, timezone: nil,
            snoozedUntil: snoozedUntil.map(NotebookReminderDate.string), installation: nil)
        _ = try await reminderMutation(payload, identity: expected)
    }
    private struct ReminderReceipt: Decodable {
        let schema_version: Int
        let mutation_id: UUID
        let reminder_id: UUID?
        let state: String
        let delivery_strategy: String?
    }
    private func reminderMutation(_ payload: NotebookReminderMutationPayload,
                                  identity expected: NotebookIdentity) async throws -> ReminderReceipt {
        guard identity() == expected else { throw NotebookFailure.identityChanged }
        do {
            let data = try await sdk.rpc("isg_notebook_reminder_mutate_v1", params: payload).execute().data
            try Task.checkCancellation()
            guard identity() == expected, data.count <= 16_384 else { throw NotebookFailure.identityChanged }
            let value = try JSONDecoder().decode(ReminderReceipt.self, from: data)
            guard value.schema_version == 1, value.mutation_id == payload.mutation else { throw NotebookFailure.invalid }
            return value
        } catch let error as PostgrestError {
            let codes = ["AUTH_REQUIRED", "FEATURE_UNAVAILABLE", "ACCESS_DENIED", "VERSION_CONFLICT",
                         "IDEMPOTENCY_CONFLICT", "VALIDATION_ERROR", "DEVICE_UNAVAILABLE"]
            if ["P0001", "28000"].contains(error.code ?? ""), codes.contains(error.message) {
                throw NotebookServerFailure(code: error.message)
            }
            throw NotebookFailure.unavailable
        }
    }
    func conflict(_ pending: NotebookPending, identity expected: NotebookIdentity) async throws -> (NotebookRecord, NotebookConflict) {
        guard identity() == expected, let wanted = pending.conflictID else { throw NotebookFailure.identityChanged }
        var after: UUID?; var version: Int64?
        for _ in 0..<50 {
            let data = try await sdk.rpc("isg_notebook_read_v1", params: ReadPayload(note: pending.intent.note, after: after)).execute().data
            try Task.checkCancellation()
            guard identity() == expected else { throw NotebookFailure.identityChanged }
            guard data.count <= 2_000_000 else { throw NotebookFailure.invalid }
            let page = try JSONDecoder().decode(NotebookConflictPage.self, from: data)
            try page.validate(noteID: pending.intent.note, after: after)
            guard !page.note.tombstone else { throw NotebookServerFailure(code: "NOTE_TOMBSTONED") }
            if let version, version != page.note.version { throw NotebookFailure.unavailable }
            version = page.note.version
            if let match = page.note.conflicts.first(where: { $0.conflict_id == wanted }) { return (page.note.record, match) }
            guard page.has_more_conflicts else { throw NotebookServerFailure(code: "CONFLICT_ALREADY_RESOLVED") }
            after = page.next_conflict_after
        }
        throw NotebookFailure.full
    }
    func identity() -> NotebookIdentity? {
        guard let session = sdk.auth.currentSession else { return nil }
        let parts = session.accessToken.split(separator: "."); guard parts.count == 3 else { return nil }
        var raw = String(parts[1]).replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        raw += String(repeating: "=", count: (4 - raw.count % 4) % 4)
        guard let data = Data(base64Encoded: raw), let claims = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let id = claims["session_id"] as? String, let sessionID = UUID(uuidString: id) else { return nil }
        return NotebookIdentity(owner: session.user.id, session: sessionID) // Correlation only; server verifies session.
    }
    private struct Payload: Codable {
        let p_mutation: UUID; let p_note: UUID; let p_action: String; let p_expected: Int64
        let p_title: String?; let p_body: String?; let p_conflict: UUID?
        enum CodingKeys: String, CodingKey { case p_mutation, p_note, p_action, p_expected, p_title, p_body, p_conflict }
        func encode(to encoder: Encoder) throws {
            var c = encoder.container(keyedBy: CodingKeys.self)
            try c.encode(p_mutation, forKey: .p_mutation); try c.encode(p_note, forKey: .p_note)
            try c.encode(p_action, forKey: .p_action); try c.encode(p_expected, forKey: .p_expected)
            if let p_title { try c.encode(p_title, forKey: .p_title) } else { try c.encodeNil(forKey: .p_title) }
            if let p_body { try c.encode(p_body, forKey: .p_body) } else { try c.encodeNil(forKey: .p_body) }
            if let p_conflict { try c.encode(p_conflict, forKey: .p_conflict) } else { try c.encodeNil(forKey: .p_conflict) }
        }
    }
    lazy var queue = NotebookQueue(storage: NotebookKeychainStorage(), current: { [weak self] in self?.identity() }, send: { [weak self] data in
        guard let self else { throw NotebookFailure.unavailable }
        do {
            if let value = try? JSONDecoder().decode(NotebookOrganizationPayload.self, from: data) {
                return try await self.sdk.rpc("isg_notebook_organize_v1", params: value).execute().data
            }
            return try await self.sdk.rpc("isg_notebook_mutate_v1", params: JSONDecoder().decode(Payload.self, from: data)).execute().data
        }
        catch let error as PostgrestError {
            let codes = ["AUTH_REQUIRED","FEATURE_UNAVAILABLE","ACCESS_DENIED","VERSION_CONFLICT","NOTE_TOMBSTONED","IDEMPOTENCY_CONFLICT","VALIDATION_ERROR","CONFLICT_ALREADY_RESOLVED"]
            if ["P0001","28000"].contains(error.code ?? ""), codes.contains(error.message) { throw NotebookServerFailure(code: error.message) }
            throw NotebookFailure.unavailable
        }
    })
}

File: android/core/data/src/main/kotlin/com/riskdetectedan/core/data/notebook/NotebookRepository.kt
package com.riskdetectedan.core.data.notebook

import com.riskdetectedan.core.data.notifications.DeviceTokenRepository
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.exceptions.RestException
import io.github.jan.supabase.postgrest.postgrest
import io.ktor.client.plugins.ResponseException
import io.ktor.client.statement.bodyAsText
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.serialization.json.*
import java.util.Base64
import java.util.UUID
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class NotebookRepository @Inject constructor(
    private val client: SupabaseClient,
    storage: NotebookEncryptedStorage,
    private val deviceTokens: DeviceTokenRepository,
) {
    val identities get() = client.auth.sessionStatus.map { identity() }.distinctUntilChanged()
    val reader = NotebookReader({ identity() }) { after ->
        client.postgrest.rpc("isg_notebook_read_v1", buildJsonObject {
            put("p_note", JsonNull); put("p_after", after?.let(::JsonPrimitive) ?: JsonNull)
        }).data
    }
    suspend fun snapshot(identity: NotebookIdentity): NotebookReader.Snapshot = reader.snapshot(identity, queue.pending(identity))
    suspend fun organization(note: String, expected: NotebookIdentity): NotebookOrganization {
        if (identity() != expected) throw NotebookFailure("IDENTITY_CHANGED")
        val raw = client.postgrest.rpc("isg_notebook_organization_v1", buildJsonObject { put("p_note", note) }).data
        currentCoroutineContext().ensureActive()
        if (identity() != expected) throw NotebookFailure("IDENTITY_CHANGED")
        require(raw.toByteArray(Charsets.UTF_8).size <= 2_000_000)
        return Json.decodeFromString<NotebookOrganization>(raw).also { it.validate(note) }
    }
    suspend fun reminders(expected: NotebookIdentity): List<NotebookReminder> {
        if (identity() != expected) throw NotebookFailure("IDENTITY_CHANGED")
        val result = mutableListOf<NotebookReminder>()
        var after: String? = null
        repeat(50) {
            val raw = client.postgrest.rpc("isg_notebook_reminders_v1", buildJsonObject {
                if (after == null) put("p_after", JsonNull) else put("p_after", after!!)
            }).data
            currentCoroutineContext().ensureActive()
            if (identity() != expected) throw NotebookFailure("IDENTITY_CHANGED")
            require(raw.toByteArray(Charsets.UTF_8).size <= 2_000_000)
            val page = Json.decodeFromString<NotebookReminderPage>(raw).also { it.validate(after) }
            result += page.reminders
            require(result.size <= 1_000)
            if (!page.has_more) return result
            after = page.next_after
        }
        throw NotebookFailure("LIMIT")
    }
    /** A reminder, optionally tied to one note, delivered by this installation's server push registration. */
    suspend fun createReminder(title: String, recurrence: String, dueAt: Instant, expected: NotebookIdentity, note: String? = null) {
        if (identity() != expected || dueAt <= Instant.now() || title.isBlank() ||
            title.codePointCount(0, title.length) > 200 || recurrence !in setOf("once", "daily", "weekly", "monthly")) {
            throw NotebookFailure("INVALID")
        }
        val zoned = dueAt.atZone(ZoneId.systemDefault())
        reminderMutation(buildJsonObject {
            put("p_mutation", UUID.randomUUID().toString()); put("p_action", "create")
            put("p_reminder", JsonNull); put("p_note", note?.let(::JsonPrimitive) ?: JsonNull); put("p_occurrence", JsonNull); put("p_expected", 0)
            put("p_title", title); put("p_recurrence", recurrence)
            put("p_local_time", zoned.format(DateTimeFormatter.ofPattern("HH:mm:ss")))
            put("p_starts_on", zoned.format(DateTimeFormatter.ISO_LOCAL_DATE)); put("p_timezone", zoned.zone.id)
            put("p_snoozed_until", JsonNull); put("p_installation", deviceTokens.installationIdForServerPush())
        }, expected).also { receipt ->
            require(receipt["state"]?.jsonPrimitive?.content == "active")
            require(receipt["delivery_strategy"]?.jsonPrimitive?.content == "server_push")
        }
    }
    suspend fun settleReminder(action: String, reminder: NotebookReminder,
                               occurrence: NotebookReminderOccurrence? = null, snoozedUntil: Instant? = null,
                               expected: NotebookIdentity) {
        require(action in setOf("complete", "snooze", "cancel") && reminder.state == "active")
        require((action == "cancel") == (occurrence == null) && (action == "snooze") == (snoozedUntil != null))
        reminderMutation(buildJsonObject {
            put("p_mutation", UUID.randomUUID().toString()); put("p_action", action); put("p_reminder", reminder.reminder_id)
            put("p_note", JsonNull)
            if (occurrence == null) put("p_occurrence", JsonNull) else put("p_occurrence", occurrence.occurrence_id)
            put("p_expected", reminder.series_version); put("p_title", JsonNull); put("p_recurrence", JsonNull)
            put("p_local_time", JsonNull); put("p_starts_on", JsonNull); put("p_timezone", JsonNull)
            if (snoozedUntil == null) put("p_snoozed_until", JsonNull) else put("p_snoozed_until", snoozedUntil.toString())
            put("p_installation", JsonNull)
        }, expected)
    }
    private suspend fun reminderMutation(arguments: JsonObject, expected: NotebookIdentity): JsonObject {
        if (identity() != expected) throw NotebookFailure("IDENTITY_CHANGED")
        try {
            val raw = client.postgrest.rpc("isg_notebook_reminder_mutate_v1", arguments).data
            currentCoroutineContext().ensureActive()
            if (identity() != expected) throw NotebookFailure("IDENTITY_CHANGED")
            require(raw.toByteArray(Charsets.UTF_8).size <= 16_384)
            return Json.parseToJsonElement(raw).jsonObject.also {
                require(it["schema_version"]?.jsonPrimitive?.int == 1)
                require(it["mutation_id"]?.jsonPrimitive?.content == arguments.getValue("p_mutation").jsonPrimitive.content)
            }
        } catch (error: Exception) {
            currentCoroutineContext().ensureActive()
            val response = when (error) { is RestException -> error.response; is ResponseException -> error.response; else -> null }
            val body = response?.let { runCatching { Json.parseToJsonElement(it.bodyAsText()).jsonObject }.getOrNull() }
            val state = body?.get("code")?.jsonPrimitive?.content
            val code = body?.get("message")?.jsonPrimitive?.content
            if (state in setOf("P0001", "28000") && code in setOf("AUTH_REQUIRED", "FEATURE_UNAVAILABLE", "ACCESS_DENIED",
                    "VERSION_CONFLICT", "IDEMPOTENCY_CONFLICT", "VALIDATION_ERROR", "DEVICE_UNAVAILABLE")) {
                throw NotebookServerFailure(code!!)
            }
            if (error is CancellationException) throw error
            throw NotebookFailure("UNAVAILABLE")
        }
    }
    suspend fun conflict(pending: NotebookPending, expected: NotebookIdentity): Pair<NotebookRecord, NotebookConflict> {
        if (identity() != expected) throw NotebookFailure("IDENTITY_CHANGED")
        val wanted = pending.conflictID ?: throw NotebookFailure("INVALID")
        var after: String? = null; var version: Long? = null
        repeat(50) {
            val raw = client.postgrest.rpc("isg_notebook_read_v1", buildJsonObject {
                put("p_note", pending.intent.note); put("p_after", after?.let(::JsonPrimitive) ?: JsonNull)
            }).data
            currentCoroutineContext().ensureActive()
            if (identity() != expected) throw NotebookFailure("IDENTITY_CHANGED")
            require(raw.toByteArray(Charsets.UTF_8).size <= 2_000_000)
            val page = Json.decodeFromString<NotebookConflictPage>(raw); page.validate(pending.intent.note, after)
            if (page.note.tombstone) throw NotebookServerFailure("NOTE_TOMBSTONED")
            if (version != null && version != page.note.version) throw NotebookFailure("UNAVAILABLE")
            version = page.note.version
            page.note.conflicts.find { it.conflict_id == wanted }?.let { return page.note.record() to it }
            if (!page.has_more_conflicts) throw NotebookServerFailure("CONFLICT_ALREADY_RESOLVED")
            after = page.next_conflict_after
        }
        throw NotebookFailure("LIMIT")
    }
    fun identity(): NotebookIdentity? = runCatching {
        val session = client.auth.currentSessionOrNull() ?: return null
        val owner = client.auth.currentUserOrNull()?.id ?: return null
        val body = Json.parseToJsonElement(String(Base64.getUrlDecoder().decode(session.accessToken.split('.')[1]), Charsets.UTF_8)).jsonObject
        NotebookIdentity(UUID.fromString(owner), UUID.fromString(body.getValue("session_id").jsonPrimitive.content))
    }.getOrNull() // Identity correlation only; PostgreSQL validates the session.
    val queue = NotebookQueue(storage, { identity() }) { args ->
        try {
            val raw = client.postgrest.rpc(if ("p_items" in args) "isg_notebook_organize_v1" else "isg_notebook_mutate_v1", args).data
            if (raw.toByteArray(Charsets.UTF_8).size > 16384) throw NotebookFailure("INVALID_RESPONSE")
            Json.parseToJsonElement(raw).jsonObject
        } catch (error: Exception) {
            currentCoroutineContext().ensureActive()
            val response = when (error) { is RestException -> error.response; is ResponseException -> error.response; else -> null }
            val body = response?.let { runCatching { Json.parseToJsonElement(it.bodyAsText()).jsonObject }.getOrNull() }
            val state = body?.get("code")?.jsonPrimitive?.content
            val code = body?.get("message")?.jsonPrimitive?.content
            if (state in setOf("P0001", "28000") && code in setOf("AUTH_REQUIRED", "FEATURE_UNAVAILABLE", "ACCESS_DENIED", "VERSION_CONFLICT", "NOTE_TOMBSTONED", "IDEMPOTENCY_CONFLICT", "VALIDATION_ERROR", "CONFLICT_ALREADY_RESOLVED")) throw NotebookServerFailure(code!!)
            throw NotebookFailure("UNAVAILABLE")
        }
    }
}
